tooling/hcpctl/pkg/kusto/client.go

```diff
@@ -0,0 +1,164 @@
+}
+
+// QueryResult represents the result of a Kusto query execution
+type QueryResult struct {
+	Data       []any
+	Columns    []Column
+	QueryStats QueryStats
+}
+
+// Column represents a column in the query result
+type Column struct {
+	Name string
+	Type string
+}
+
+// QueryStats represents statistics about the query execution
+type QueryStats struct {
+	ExecutionTime time.Duration
+	TotalRows     int
+	DataSize      int64
+}
+
+// NewClient creates a new Azure Data Explorer client
+func NewClient(endpoint string) (*Client, error) {
+	if endpoint == "" {
+		return nil, fmt.Errorf("cluster name is required")
+	}
+
+	// Create connection string builder
+	kcsb := kusto.NewConnectionStringBuilder(endpoint)
+
+	// Use Azure default credential chain for authentication
+	kcsb = kcsb.WithDefaultAzureCredential()
+
+	// Create Kusto client with authentication
+	kustoClient, err := kusto.New(kcsb)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kusto client: %w", err)
+	}
+
+	return &Client{
+		Endpoint:    endpoint,
+		kustoClient: kustoClient,
+	}, nil
+}
+
+// ExecutePreconfiguredQuery executes a KQL query against the Azure Data Explorer cluster
+// This is to execute queries bundled with this package. Not to be used for custom queries.
+func (c *Client) ExecutePreconfiguredQuery(ctx context.Context, query *ConfigurableQuery, rowType any, timeout time.Duration) (*QueryResult, error) {
+	// Create a context with timeout
+	queryCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	fmt.Printf("Executing query: %s\n", query.Name)
+
+	// Execute the query using kql.Builder with AddUnsafe for dynamic queries
+	// This allows us to use dynamic query strings
+	iter, err := c.kustoClient.Query(queryCtx, query.Database, query.Query, kusto.QueryParameters(query.Parameters))
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute query: %w", err)
+	}
+	defer iter.Stop()
+
+	// Process results
+	var data []any
+	var columns []Column
+	var totalRows int
+	var dataSize int64
+	startTime := time.Now()
+
+	// Process rows using DoOnRowOrError
+	err = iter.DoOnRowOrError(func(row *table.Row, e *kustoErrors.Error) error {
+		if e != nil {
+			return fmt.Errorf("failed to process row: %w", e)
+		}
+
+		// Extract column information from the first row
+		if totalRows == 0 {
+			// Get column information from the row
+			colNames := row.ColumnNames()
+			colTypes := row.ColumnTypes
+			for i, name := range colNames {
+				columns = append(columns, Column{
+					Name: name,
+					Type: string(colTypes[i].Type),
+				})
+			}
+		}
+
+		// Convert row to map
+		rowData := reflect.New(reflect.TypeOf(rowType)).Interface()
+		err := row.ToStruct(rowData)
+		if err != nil {
+			return fmt.Errorf("failed to convert row to struct: %w", err)
+		}
+		// from the docs https://pkg.go.dev/github.com/Azure/azure-kusto-go#readme-query-into-structs
+		if row.Replace {
+			data = data[:0]
+		}
+		data = append(data, rowData)
```

Comment:
As a starting point, by all means let's do this.  Our next iteration needs to take this an instead of holding all the data in memory, we need to stream it out to disk.